File: advent_of_code/2022/05/day05.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Stack struct {
	values []string
}

func NewStack() *Stack {
	s := Stack{}
	return &s
}

func (s *Stack) push(v string) {
	s.values = append(s.values, v)
}

func (s *Stack) pop() *string {
	if len(s.values) == 0 {
		return nil
	}
	n := len(s.values)
	v := s.values[n-1]
	s.values = s.values[0 : n-1]
	return &v
}

func (s *Stack) peek() *string {
	if len(s.values) == 0 {
		return nil
	}
	return &s.values[len(s.values)-1]
}

func (stack *Stack) str() string {
	s := "["
	for i := 0; i < len(stack.values); i++ {
		if i != 0 {
			s += ", "
		}
		s += stack.values[i]
	}
	s += "]"
	return s
}

func (stack *Stack) size() int {
	return len(stack.values)
}

func dumpStacks(stacks []Stack) {
	for i := 0; i < len(stacks); i++ {
		fmt.Printf("%d: %s\n", i+1, stacks[i].str())
	}
}

func part1(input string) string {
	lines := strings.Split(input, "\n")
	n := 0
	for len(lines[n]) != 0 {
		n++
	}

	nStacks := (len(lines[n-1]) + 1) / 4
	stacks := make([]Stack, nStacks)
	for i := 0; i < nStacks; i++ {
		stacks[i] = *NewStack()
	}

	for i := n - 2; i >= 0; i-- {
		for j := 0; j < nStacks; j++ {
			c := string(lines[i][j*4+1])
			if c != " " {
				stacks[j].push(c)
			}
		}
	}

	// dumpStacks(stacks)

	for i := n + 1; i < len(lines); i++ {
		move := strings.Split(lines[i], " ")
		n, _ := strconv.Atoi(move[1])
		from, _ := strconv.Atoi(move[3])
		to, _ := strconv.Atoi(move[5])
		from -= 1
		to -= 1

		for j := 0; j < n; j++ {
			stacks[to].push(*stacks[from].pop())
		}
	}

	res := ""
	for _, stack := range stacks {
		res += *stack.peek()
	}
	return res
}

func part2(input string) string {
	lines := strings.Split(input, "\n")
	n := 0
	for len(lines[n]) != 0 {
		n++
	}

	nStacks := (len(lines[n-1]) + 1) / 4
	stacks := make([]Stack, nStacks)
	for i := 0; i < nStacks; i++ {
		stacks[i] = *NewStack()
	}

	for i := n - 2; i >= 0; i-- {
		for j := 0; j < nStacks; j++ {
			c := string(lines[i][j*4+1])
			if c != " " {
				stacks[j].push(c)
			}
		}
	}

	temp := *NewStack()
	for i := n + 1; i < len(lines); i++ {
		move := strings.Split(lines[i], " ")
		n, _ := strconv.Atoi(move[1])
		from, _ := strconv.Atoi(move[3])
		to, _ := strconv.Atoi(move[5])
		from -= 1
		to -= 1

		for j := 0; j < n; j++ {
			temp.push(*stacks[from].pop())
		}
		for j := 0; j < n; j++ {
			stacks[to].push(*temp.pop())
		}
	}

	res := ""
	for _, stack := range stacks {
		res += *stack.peek()
	}
	return res
}

const test_input = `    [D]    
[N] [C]    
[Z] [M] [P]
 1   2   3 

move 1 from 2 to 1
move 3 from 1 to 3
move 2 from 2 to 1
move 1 from 1 to 2`

func main() {
	const DEBUG = true
	filename := "input.txt"
	var input string
	if DEBUG {
		input = test_input
	} else {
		s, err := os.ReadFile(filename)
		if err != nil {
			panic(err)
		}
		input = string(s)
	}

	fmt.Printf("2022 Day 5, part 1: %v\n", part1(input))
	fmt.Printf("2022 Day 5, part 2: %v\n", part2(input))
}

File: scala/aoc/src/main/scala/aoc2023/day05.scala
package aoc2023

import scala.concurrent.{Future, Await, ExecutionContext}
import scala.collection.immutable.NumericRange
import scala.math.min
import scala.concurrent.duration.Duration
import scala.collection.mutable.ArrayBuffer

val testInput = """
seeds: 79 14 55 13

seed-to-soil map:
50 98 2
52 50 48

soil-to-fertilizer map:
0 15 37
37 52 2
39 0 15

fertilizer-to-water map:
49 53 8
0 11 42
42 0 7
57 7 4

water-to-light map:
88 18 7
18 25 70

light-to-temperature map:
45 77 23
81 45 19
68 64 13

temperature-to-humidity map:
0 69 1
1 0 69

humidity-to-location map:
60 56 37
56 93 4
"""

case class RangeMap(source: NumericRange[Long], dest: NumericRange[Long]):
  override def toString(): String =
    s"source=$source, dest=$dest"
  def contains(i: Long) = source.contains(i)
  def transform(i: Long) =
    if contains(i)
    then i - source.start + dest.start
    else i

case class ResourceMap(source: String, dest: String, ranges: Array[RangeMap]):
  override def toString(): String =
    s"$source-to-$dest:\n${ranges.map(_.toString()).mkString("\n")}"

def readInput(input: String): (Array[Long], Array[ResourceMap]) =
  val configs = input.trim().split("\n\n").map(_.trim()).filter(_.nonEmpty)
  val seeds = configs.head.stripPrefix("seeds: ").split("\\s+").map(_.toLong)
  val resources = configs.tail.map:
    config =>
      val lines = config.split("\n").map(_.trim()).filter(_.nonEmpty)
      val Array(sourceId, _, destId) = lines.head.stripSuffix(" map:").split("-")
      val rangeMaps = lines.tail.map:
        line =>
          val Array(dest, source, len) = line.split("\\s+").map(_.toLong)
          RangeMap((source until (source + len)), (dest until (dest + len)))
      ResourceMap(sourceId, destId, rangeMaps)
  (seeds, resources)


def part1(input: String): Long =
  val (startSeeds, resources) = readInput(input)
  resources.foldLeft(startSeeds):
    case (seeds, ResourceMap(_, _, ranges)) =>
      seeds.map:
        seed =>
          ranges.find(range => range.contains(seed)) match
            case Some(range) => range.transform(seed)
            case None => seed
  .min


// Concurrent solution; runs real slow
def part2_concurrent(input: String): Long =
  val (startSeeds, resources) = readInput(input)
  val seedRanges = startSeeds.toList.grouped(2).toList.map(vs => vs(0) until vs(0) + vs(1)).toList

  given ExecutionContext = ExecutionContext.global

  def divideRange(startRange: NumericRange[Long], chunkSize: Int): List[NumericRange[Long]] =
    var range = startRange
    val res = ArrayBuffer[NumericRange[Long]]()
    while range.nonEmpty do
      val (left, right) = range.splitAt(chunkSize)
      res += left
      range = right
    res.toList

  def process(seedRange: NumericRange[Long]): Future[Long] =
    Future:
      seedRange.map:
        seed =>
          resources.foldLeft(seed):
            case (currSeed, ResourceMap(_, _, ranges)) =>
              ranges.find(range => range.contains(currSeed)) match
                case Some(range) => range.transform(currSeed)
                case None => currSeed
      .min

  var res = Long.MaxValue
  val chunkSize = 100_000
  for seedRange <- seedRanges do
    val f = Future.sequence:
      divideRange(seedRange, chunkSize).map:
        range =>
          process(range)
    val temp = Await.result(f, Duration.Inf)
    res = min(res, temp.min)
    
  res


extension(range: NumericRange[Long])
  def contains(o: NumericRange[Long]) = range.start <= o.start && o.end <= range.end
  def overlaps(o: NumericRange[Long]) = range.contains(o.start) || range.contains(o.end)
  // def cut(o: NumericRange[Long]) =
  //   if range == o then range
  //   else if range.start == o.start
  //     if range.end < o.end then
  //       ???
    


def part2(input: String): Long =
  val (startSeeds, resources) = readInput(input)

  0L


@main def main() =
  val debug = false
  val input = if debug
    then testInput
    else io.Source.fromFile("input.txt").getLines().mkString("\n")
  println(s"2023 Day 5, Part 1: ${part1(input)}")
  println(s"2023 Day 5, Part 2: ${part2(input)}")
